TSPL.docc/ReferenceManual/Types.md

```diff
@@ -895,6 +904,78 @@ could return a value of type `T` or `Dictionary<String, T>`.
 >
 > *opaque-type* → **`some`** *type*
 
+## Boxed Protocol Type
+
+A *boxed protocol type* defines a type
+that conforms to a protocol or protocol composition,
+with the ability for that conforming type
+to vary while the program is running.
+
+Boxed protocol types have the following form:
+
+```swift
+any <#constraint#>
+```
+
+The *constraint* is a protocol type,
+protocol composition type,
+a metatype of a protocol type,
+or a metatype of a protocol composition type.
+
+At runtime,
+an instance of a boxed protocol type can contain a value
+of any type that satisfies the *constraint*.
+This behavior contrasts with how an opaque types work,
+where there is some specific conforming type known at compile time.
+The additional level of indirection that's used
+when working with a boxed protocol type is called :newTerm:`boxing`.
+Boxing typically requires a separate memory allocation for storage
+and an additional level of indirection for access,
+which incurs a performance cost at runtime.
+
+Applying `any` to the `Any` or `AnyObject` types
+is redundant and has no effect.
+
+<!--
+  - test: `any-any-does-nothing`
+
+   >> var x: any Any = 12
+   >> var y: Any = 12
+   >> print(type(of: x))
+   << Int
+   >> print(type(of: y))
+   << Int
+   >> print(type(of: x) == type(of: y))
+   << true
+-->
+
+<!--
+  - test: `any-anyobject-does-nothing`
+
+   >> import Foundation
+   >> var x: any AnyObject = NSNumber(value: 12)
+   >> var y: AnyObject = NSNumber(value: 12)
+   >> print(type(of: x))
+   << __NSCFNumber
+   >> print(type(of: y))
+   << __NSCFNumber
+   >> print(type(of: x) == type(of: y))
+   << true
+-->
+
+<!--
+Contrast P.Type with (any P.Type) and (any P).Type
+https://github.com/apple/swift-evolution/blob/main/proposals/0335-existential-any.md#metatypes
+-->
+ 
+
+```
+Grammar of a boxed protocol type
+
+boxed-protocol-type --> ``any`` type
```

Comment:
Q: would `Any` or `AnyObject` be considered a boxed protocol type for the purposes of the formal grammar, and if so does that need to be reflected here?

Reply:
I don't think `Any` or `AnyObject` are considered boxed protocol types — just that writing `any` before them has no effect.

@hborla Can you confirm?

Reply:
`Any` and `AnyObject` are considered "boxed protocol types". I wouldn't think of it as "`any` before them has no effect", but rather that the `any` keyword can be omitted. The `any` keyword is still sometimes necessary for metatypes of `Any` and `AnyObject`, because `(any Any).Type` (previously `Any.Protocol`) and `Any.Type` are different types.

Technically, `Any` is a keyword, so its grammar is special. `AnyObject` isn't grammatically different than other types. I would assume the `type` grammar covers these already though.

Reply:
Right, but read strictly this grammar does not describe the permitted omission of `any` before `AnyObject`.

Reply:
Oh, I see what you're saying, but `Any` and `AnyObject` are not determined to be existential types by the grammar; that's a semantic decision, and there are certain contexts where `Any` and `AnyObject` are requirements instead of existential types. I think that special rule about omission of `any` should be described in the prose and not in the production rules in the grammar. SE-0335 does the same thing

Reply:
@hborla

Reply:
Calling this rule out in prose, how about changing this text:

> Applying `any` to the `Any` or `AnyObject` types is redundant and has no effect.

To instead read:

> Applying `any` to the `Any` or `AnyObject` types has no effect, because those types are already boxed protocol types.

https://github.com/apple/swift-book/pull/103/files/f7c0c8181dda3ab2ec9c8a9dd32ddf9d6bdd74e0#diff-2336c9bd1d181f9592d46cfd0d8b51a0db4ed057016a8f0a7f3d26baab0eff9cR936

Reply:
Changed in f40066e0.